libdrgn/arch_x86_64.c

```diff
@@ -623,14 +654,14 @@ linux_kernel_pgtable_iterator_next_x86_64(struct drgn_program *prog,
 			uint64_t entry = it->table[level][it->index[level]++];
 			if (bswap)
 				entry = bswap_64(entry);
-			table = entry & ADDRESS_MASK;
-			if (!(entry & PRESENT) || (entry & PSE) || level == 0) {
+			table = entry & PTE_PFN_MASK;
+			if (!(entry & _PAGE_PRESENT) || (entry & PSE) || level == 0) {
 				uint64_t mask = (UINT64_C(1) <<
 						 (PAGE_SHIFT +
 						  PGTABLE_SHIFT * level)) - 1;
 				*virt_addr_ret = virt_addr & ~mask;
-				if (entry & PRESENT)
-					*phys_addr_ret = table & ~mask;
+				if (pte_present(entry))
+					*phys_addr_ret = pte_phys(entry);
```

Comment:
This will also get entries for higher levels (for huge pages or unmapped levels), so the `pte_*` names are not completely accurate. The functions themselves probably also need to depend on the level as noted above.